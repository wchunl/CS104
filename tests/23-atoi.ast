ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_FUNCTION "" (7.5.1)
|   |   TOK_TYPE_ID "" (7.5.1)
|   |   |   TOK_INT "int" (7.5.1)
|   |   |   TOK_IDENT "atoi" (7.5.5)
|   |   TOK_PARAM "(" (7.5.10)
|   |   |   TOK_TYPE_ID "" (7.5.11)
|   |   |   |   TOK_STRING "string" (7.5.11)
|   |   |   |   TOK_IDENT "str" (7.5.18)
|   |   TOK_BLOCK "{" (7.5.23)
|   |   |   TOK_FUNCTION "" (7.6.12)
|   |   |   |   TOK_VARDECL "=" (7.6.12)
|   |   |   |   |   TOK_INT "int" (7.6.4)
|   |   |   |   |   |   TOK_TYPE_ID "neg" (7.6.8)
|   |   |   |   |   TOK_INTCON "0" (7.6.14)
|   |   |   |   TOK_VARDECL "=" (7.7.12)
|   |   |   |   |   TOK_INT "int" (7.7.4)
|   |   |   |   |   |   TOK_TYPE_ID "num" (7.7.8)
|   |   |   |   |   TOK_INTCON "0" (7.7.14)
|   |   |   |   TOK_VARDECL "=" (7.8.14)
|   |   |   |   |   TOK_INT "int" (7.8.4)
|   |   |   |   |   |   TOK_TYPE_ID "digit" (7.8.8)
|   |   |   |   |   TOK_INTCON "0" (7.8.16)
|   |   |   |   TOK_BLOCK "{" (7.9.4)
|   |   |   |   |   TOK_FUNCTION "" (7.9.5)
|   |   |   |   |   |   TOK_IF "if" (7.9.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.9.9)
|   |   |   |   |   |   |   |   TOK_NE "!=" (7.9.18)
|   |   |   |   |   |   |   |   |   TOK_IDENT "str" (7.9.14)
|   |   |   |   |   |   |   |   |   TOK_NULLPTR "nullptr" (7.9.21)
|   |   |   |   |   |   |   TOK_CALL "(" (7.9.36)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.9.31)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""str != nullptr"" (7.9.37)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""23-atoi.oc"" (7.9.55)
|   |   |   |   |   |   |   |   |   TOK_INTCON "9" (7.9.69)
|   |   |   |   ';' ";" (7.9.73)
|   |   |   |   TOK_IF "if" (7.10.4)
|   |   |   |   |   TOK_NE "!=" (7.10.15)
|   |   |   |   |   |   TOK_INDEX "[" (7.10.11)
|   |   |   |   |   |   |   TOK_IDENT "str" (7.10.8)
|   |   |   |   |   |   |   TOK_INTCON "0" (7.10.12)
|   |   |   |   |   |   TOK_CHARCON "'\0'" (7.10.18)
|   |   |   |   |   TOK_BLOCK "{" (7.10.24)
|   |   |   |   |   |   TOK_FUNCTION "" (7.11.7)
|   |   |   |   |   |   |   TOK_IF "if" (7.11.7)
|   |   |   |   |   |   |   |   TOK_EQ "==" (7.11.18)
|   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.11.14)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "str" (7.11.11)
|   |   |   |   |   |   |   |   |   |   TOK_INTCON "0" (7.11.15)
|   |   |   |   |   |   |   |   |   TOK_CHARCON "'-'" (7.11.21)
|   |   |   |   |   |   |   |   TOK_BLOCK "{" (7.11.26)
|   |   |   |   |   |   |   |   |   TOK_FUNCTION "" (7.12.16)
|   |   |   |   |   |   |   |   |   |   '=' "=" (7.12.16)
|   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.12.10)
|   |   |   |   |   |   |   |   |   |   |   '+' "+" (7.12.24)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.12.18)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.12.26)
|   |   |   |   |   |   |   |   |   |   '=' "=" (7.13.14)
|   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "neg" (7.13.10)
|   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.13.16)
|   |   |   |   |   |   |   TOK_VARDECL "=" (7.15.18)
|   |   |   |   |   |   |   |   TOK_INT "int" (7.15.7)
|   |   |   |   |   |   |   |   |   TOK_TYPE_ID "contin" (7.15.11)
|   |   |   |   |   |   |   |   TOK_INTCON "1" (7.15.20)
|   |   |   |   |   |   |   TOK_WHILE "while" (7.16.7)
|   |   |   |   |   |   |   |   TOK_IDENT "contin" (7.16.14)
|   |   |   |   |   |   |   |   TOK_BLOCK "{" (7.16.22)
|   |   |   |   |   |   |   |   |   TOK_FUNCTION "" (7.17.10)
|   |   |   |   |   |   |   |   |   |   TOK_IF "if" (7.17.10)
|   |   |   |   |   |   |   |   |   |   |   TOK_EQ "==" (7.17.25)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.17.17)
|   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "str" (7.17.14)
|   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.17.18)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_CHARCON "'\0'" (7.17.28)
|   |   |   |   |   |   |   |   |   |   |   TOK_BLOCK "{" (7.17.34)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_FUNCTION "" (7.18.20)
|   |   |   |   |   |   |   |   |   |   |   |   |   '=' "=" (7.18.20)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "contin" (7.18.13)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "0" (7.18.22)
|   |   |   |   |   |   |   |   |   |   |   TOK_IFELSE "{" (7.19.16)
|   |   |   |   |   |   |   |   |   |   |   |   TOK_FUNCTION "" (7.20.21)
|   |   |   |   |   |   |   |   |   |   |   |   |   TOK_VARDECL "=" (7.20.21)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INT "int" (7.20.13)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_TYPE_ID "chr" (7.20.17)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.20.26)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "str" (7.20.23)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.20.27)
|   |   |   |   |   |   |   |   |   |   |   |   |   '=' "=" (7.21.19)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.21.13)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   '+' "+" (7.21.27)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "digit" (7.21.21)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.21.29)
|   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IF "if" (7.22.13)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   '<' "<" (7.22.21)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "chr" (7.22.17)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_CHARCON "'0'" (7.22.23)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   '=' "=" (7.22.35)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "contin" (7.22.28)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "0" (7.22.37)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IFELSE "if" (7.23.18)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   '>' ">" (7.23.26)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "chr" (7.23.22)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_CHARCON "'9'" (7.23.28)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   '=' "=" (7.23.40)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "contin" (7.23.33)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "0" (7.23.42)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IFELSE "=" (7.24.22)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "num" (7.24.18)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   '-' "-" (7.24.39)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   '+' "+" (7.24.33)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   '*' "*" (7.24.28)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "num" (7.24.24)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_INTCON "10" (7.24.30)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "chr" (7.24.35)
|   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   |   TOK_CHARCON "'0'" (7.24.41)
|   |   |   |   |   |   |   TOK_IF "if" (7.27.7)
|   |   |   |   |   |   |   |   TOK_IDENT "neg" (7.27.11)
|   |   |   |   |   |   |   |   '=' "=" (7.27.20)
|   |   |   |   |   |   |   |   |   TOK_IDENT "num" (7.27.16)
|   |   |   |   |   |   |   |   |   '-' "-" (7.27.22)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "num" (7.27.24)
|   |   |   |   TOK_RETURN "return" (7.29.4)
|   |   |   |   |   TOK_IDENT "num" (7.29.11)
|   TOK_FUNCTION "" (7.32.1)
|   |   TOK_TYPE_ID "" (7.32.1)
|   |   |   TOK_INT "int" (7.32.1)
|   |   |   TOK_IDENT "main" (7.32.5)
|   |   TOK_PARAM "(" (7.32.10)
|   |   |   TOK_TYPE_ID "" (7.32.11)
|   |   |   |   TOK_INT "int" (7.32.11)
|   |   |   |   TOK_IDENT "argc" (7.32.15)
|   |   |   TOK_TYPE_ID "" (7.32.21)
|   |   |   |   TOK_ARRAY "array" (7.32.21)
|   |   |   |   |   TOK_TYPE_ID "string" (7.32.27)
|   |   |   |   TOK_IDENT "argv" (7.32.35)
|   |   TOK_BLOCK "{" (7.32.41)
|   |   |   TOK_FUNCTION "" (7.33.13)
|   |   |   |   TOK_VARDECL "=" (7.33.13)
|   |   |   |   |   TOK_INT "int" (7.33.4)
|   |   |   |   |   |   TOK_TYPE_ID "argi" (7.33.8)
|   |   |   |   |   TOK_INTCON "1" (7.33.15)
|   |   |   |   TOK_VARDECL "=" (7.34.15)
|   |   |   |   |   TOK_STRING "string" (7.34.4)
|   |   |   |   |   |   TOK_TYPE_ID "arg" (7.34.11)
|   |   |   |   |   TOK_NULLPTR "nullptr" (7.34.17)
|   |   |   |   TOK_WHILE "while" (7.35.4)
|   |   |   |   |   '<' "<" (7.35.16)
|   |   |   |   |   |   TOK_IDENT "argi" (7.35.11)
|   |   |   |   |   |   TOK_IDENT "argc" (7.35.18)
|   |   |   |   |   TOK_BLOCK "{" (7.35.24)
|   |   |   |   |   |   TOK_FUNCTION "" (7.36.11)
|   |   |   |   |   |   |   '=' "=" (7.36.11)
|   |   |   |   |   |   |   |   TOK_IDENT "arg" (7.36.7)
|   |   |   |   |   |   |   |   TOK_INDEX "[" (7.36.17)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.36.13)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.36.18)
|   |   |   |   |   |   |   TOK_CALL "(" (7.37.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putstr" (7.37.7)
|   |   |   |   |   |   |   |   TOK_IDENT "arg" (7.37.15)
|   |   |   |   |   |   |   TOK_CALL "(" (7.38.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putstr" (7.38.7)
|   |   |   |   |   |   |   |   TOK_STRINGCON "" = "" (7.38.15)
|   |   |   |   |   |   |   TOK_CALL "(" (7.39.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putint" (7.39.7)
|   |   |   |   |   |   |   |   TOK_CALL "(" (7.39.20)
|   |   |   |   |   |   |   |   |   TOK_IDENT "atoi" (7.39.15)
|   |   |   |   |   |   |   |   |   TOK_IDENT "arg" (7.39.21)
|   |   |   |   |   |   |   TOK_CALL "(" (7.40.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putchr" (7.40.7)
|   |   |   |   |   |   |   |   TOK_CHARCON "'\n'" (7.40.15)
|   |   |   |   |   |   |   '=' "=" (7.41.12)
|   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.41.7)
|   |   |   |   |   |   |   |   '+' "+" (7.41.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.41.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.41.21)
|   |   |   |   TOK_RETURN "return" (7.43.4)
|   |   |   |   |   TOK_INTCON "0" (7.43.11)
