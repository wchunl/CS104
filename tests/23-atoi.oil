fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
atoi:     .function int
          .param ptr str
          .local int neg
          .local int num
          .local int digit
          neg = 0
          num = 0
          digit = 0
.if0:     $t0:i = str != nullptr
          $t1:i = not $t0:i
          goto .el0 if not $t1:i
.th0:     $t2:i = "str != nullptr"
          fail($t2:i)
.fi0:     
.if1:     $t3:i = str[0 * :p]
          $t4:i = $t3:i != '\0'
          goto .el1 if not $t4:i
.th1:     
.if2:     $t5:i = str[0 * :p]
          $t6:i = $t5:i == '-'
          goto .el2 if not $t6:i
.th2:     $t7:i = digit + 1
          digit = $t7:i
          neg = 1
.fi2:     contin = 1
.wh0:     $t8:i = contin
          goto .od0 if not $t8:i
.do0:     
.if3:     $t9:i = str[digit * :p]
          $t10:i = $t9:i == '\0'
          goto .el3 if not $t10:i
.th3:     contin = 0
.el3:     $t11:p = str[digit * :p]
          chr = $t11:i
          $t12:i = digit + 1
          digit = $t12:i
.if4:     $t13:i = chr < '0'
          goto .el4 if not $t13:i
.th4:     contin = 0
.el4:     
.if5:     $t14:i = chr > '9'
          goto .el5 if not $t14:i
.th5:     contin = 0
.el5:     $t15:i = + - '0'
          num = $t15:i
.fi5:     
.fi4:     
.fi3:     
.od0:     
.if6:     $t16:i = neg
          goto .el6 if not $t16:i
.th6:     $t17:i = - num
          num = $t17:i
.fi6:     
.fi1:     return num
          return 
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local int argi
          .local ptr arg
          argi = 1
          arg = nullptr
.wh1:     $t18:i = argi < argc
          goto .od1 if not $t18:i
.do1:     $t19:p = argv[argi * :p]
          arg = $t19:i
          putstr(arg)
          putstr(" = ")
          atoi(arg)
          putint($t20:i)
          putchr('\n')
          $t21:i = argi + 1
          argi = $t21:i
.od1:     return 0
          return 
          .end
