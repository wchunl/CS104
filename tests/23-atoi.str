string_set[   2]:    2594088158685378223 0xcec9e8->"10"
                   12597418120160561823 0xce7b18->"int"
string_set[   3]:   11752130776608228127 0xcea748->"'\0'"
string_set[   5]:    1370730858159036685 0xceac38->"+"
string_set[   8]:   13497578857207247412 0xce79d8->"string"
string_set[   9]:   10319632053171904911 0xceb108->"while"
string_set[  10]:    9201800968376702500 0xce8c08->"getchr"
string_set[  12]:   14643508047410943861 0xced2c8->"argv"
                    1902468552727064667 0xce99b8->"digit"
string_set[  14]:   15816506932648033945 0xce3218->"("
                    1396094175042499165 0xcebc48->"<"
string_set[  15]:   12868445110721718657 0xce7ce8->")"
string_set[  17]:    6142509188972423790 0xce3558->""
string_set[  20]:    1583576014053361350 0xce8858->"str"
string_set[  21]:    9317282556992397945 0xce9ba8->"if"
string_set[  22]:   10159970873491820195 0xceae88->"1"
string_set[  26]:   12291392672316406109 0xce7e78->"line"
                    9611428142580219387 0xce3578->"fail"
                   11597697714117577063 0xce9878->"="
string_set[  28]:   18413295556186398326 0xce80c8->"putint"
string_set[  29]:   11911796782241914524 0xce7648->"void"
string_set[  33]:    6814028903552918164 0xced168->"array"
string_set[  35]:   11868120863286193964 0xcecf38->"main"
string_set[  41]:   11038708016986559757 0xcec5e8->"'9'"
                   12051031859636400188 0xcea728->"]"
string_set[  42]:   13686721134187162920 0xcea3e8->"}"
string_set[  43]:   14849783028996024063 0xce9d78->"not"
string_set[  45]:    1145006563990091304 0xce8008->"neg"
string_set[  49]:   12679825168387489523 0xce86d8->"num"
string_set[  50]:      21606623943142546 0xce9608->"{"
string_set[  52]:    5344662657899890615 0xce7e98->";"
string_set[  54]:    7296690267450279874 0xce8338->"chr"
string_set[  55]:    2297668033614959926 0xce9918->"0"
string_set[  56]:    3701972582333163920 0xcecef8->"argc"
string_set[  59]:    6192799629375248351 0xcedd68->"'\n'"
string_set[  60]:    1873634190020599143 0xceb958->"else"
string_set[  63]:   14526079891305879776 0xce96a8->">"
                    6378628904606585349 0xce9288->"atoi"
string_set[  64]:   16579916199511856337 0xce9068->"getln"
string_set[  67]:    4706206694200543916 0xce7bd8->","
                   12808232827189627082 0xcea0c8->"nullptr"
string_set[  69]:    5266873771696973564 0xced058->"argi"
string_set[  72]:    7869624569597532036 0xcea348->""23-atoi.oc""
string_set[  73]:    9386170080689500517 0xce7aa8->"expr"
string_set[  76]:   11166701219533395393 0xce8e28->"getstr"
string_set[  78]:    7761470112706414413 0xced668->"arg"
string_set[  80]:    7211130878751977441 0xce81f8->"putchr"
                    2617056156435215158 0xcea028->"!="
string_set[  83]:   13581035385482130682 0xcea668->"["
string_set[  84]:    5720712447182791414 0xcec058->"'0'"
string_set[  85]:    7372855250431704371 0xceca88->"-"
                    6006759838892838694 0xceab98->"'-'"
                    4017852018407011111 0xce7cc8->"file"
string_set[  86]:   17605730733847049629 0xceb148->"contin"
string_set[  87]:   14635384904416252381 0xcece48->"return"
string_set[  89]:    6637313742931709005 0xcec688->"*"
                   13348923631108115385 0xce84b8->"putstr"
                   13599092719175287697 0xcea288->""str != nullptr""
                    6623666755989985924 0xcea408->"9"
string_set[  91]:    5090247647583737318 0xced588->"" = ""
string_set[  95]:   14736352883882276958 0xcea888->"=="
load_factor = 0.629
bucket_count = 97
max_bucket_size = 4
