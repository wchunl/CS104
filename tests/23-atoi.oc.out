; 0.  1: # 1 "23-atoi.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "23-atoi.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "23-atoi.oc" 2
; 7.  4: 
; 7.  5: int atoi (string str) {
; 7.  6:    int neg = 0;
; 7.  7:    int num = 0;
; 7.  8:    int digit = 0;
; 7.  9:    {if (not (str != nullptr)) fail ("str != nullptr", "23-atoi.oc", 9);};
; 7. 10:    if (str[0] != '\0') {
; 7. 11:       if (str[0] == '-') {
; 7. 12:          digit = digit + 1;
; 7. 13:          neg = 1;
; 7. 14:       }
; 7. 15:       int contin = 1;
; 7. 16:       while (contin) {
; 7. 17:          if (str[digit] == '\0') {
; 7. 18:             contin = 0;
; 7. 19:          }else {
; 7. 20:             int chr = str[digit];
; 7. 21:             digit = digit + 1;
; 7. 22:             if (chr < '0') contin = 0;
; 7. 23:             else if (chr > '9') contin = 0;
; 7. 24:             else num = num * 10 + chr - '0';
; 7. 25:          }
; 7. 26:       }
; 7. 27:       if (neg) num = - num;
; 7. 28:    }
; 7. 29:    return num;
; 7. 30: }
; 7. 31: 
; 7. 32: int main (int argc, array<string> argv) {
; 7. 33:    int argi = 1;
; 7. 34:    string arg = nullptr;
; 7. 35:    while (argi < argc) {
; 7. 36:       arg = argv[argi];
; 7. 37:       putstr (arg);
; 7. 38:       putstr (" = ");
; 7. 39:       putint (atoi (arg));
; 7. 40:       putchr ('\n');
; 7. 41:       argi = argi + 1;
; 7. 42:    }
; 7. 43:    return 0;
; 7. 44: }
hello